parquet/src/column/writer/mod.rs

```diff
@@ -1152,9 +1213,78 @@ fn compare_greater_byte_array_decimals(a: &[u8], b: &[u8]) -> bool {
     (a[1..]) > (b[1..])
 }
 
+/// Truncate a UTF8 slice to the longest prefix that is still a valid UTF8 string, while being less than `length` bytes.
+fn truncate_utf8(data: &str, length: usize) -> Option<Vec<u8>> {
+    // We return values like that at an earlier stage in the process.
+    assert!(data.len() >= length);
+    let mut char_indices = data.char_indices();
+
+    // We know `data` is a valid UTF8 encoded string, which means it has at least one valid UTF8 byte, which will make this loop exist.
+    while let Some((idx, c)) = char_indices.next_back() {
+        let split_point = idx + c.len_utf8();
+        if split_point <= length {
+            return data.as_bytes()[0..split_point].to_vec().into();
+        }
+    }
+
+    unreachable!()
```

Comment:
```suggestion
    None
```
I think this is reachable.

Consider `data` containing a single character with a 3 byte encoding, and a `length` of `1`. `data.len() >= length` but the loop will fail to find a character

Reply:
I was thinking about this today but was drawing a blank when trying to formulate at test case for this issue. I think I originally assumed that any UTF8 character has a valid sub-character, which is wrong. Also added a test case to validate that we're handling this case correctly.